File: source/downloads/code/LeetCode/Java/Search-for-A-Range.java
public class Solution {
    public int[] searchRange(int[] nums, int target) {
        int n = nums.length;
        int[] ret = {-1, -1};
        if(n==0) return ret;
        int l = 0;
        int r = n-1;

        while(l<=r){
            int mid = l+(r-l)/2;
            if(nums[mid]<target)
                l = mid+1;
            else
                r = mid-1;
        }
        // 有可能找不到
        if(l>=n || nums[l]!=target) return ret;
        ret[0] = l;

        l=0;
        r = n-1;
        while(l<=r){
            int mid = l+(r-l)/2;
            if(nums[mid]<=target)
                l = mid+1;
            else
                r = mid-1;
        }
        ret[1] = r;
        return ret;
    }
}

File: downloads/code/LeetCode/Python/Search-for-a-Range.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if not nums:
            return [-1,-1]
        ret = [-1,-1]
        start = 0
        end = len(nums)-1

        while start <= end:
            mid = (start+end)/2
            if target == nums[mid]:
                break
            elif target > nums[mid]:
                start = mid+1
            else:
                end = mid-1
        # Not found
        if start > end:
            return ret

        l = mid
        r = mid
	# Find range       
	while l>0 and  nums[mid] == nums[l-1]:
            l -= 1
        while r < end and nums[mid] == nums[r+1]:
            r += 1
        return [l,r]

